client/app/pages/queries/QuerySource.jsx

```diff
@@ -1,19 +1,123 @@
-import React from 'react';
+import { find, map, clone } from 'lodash';
+import React, { useState, useRef, useEffect, useMemo, useCallback } from 'react';
 import PropTypes from 'prop-types';
 import { react2angular } from 'react2angular';
+import Select from 'antd/lib/select';
 import { routesToAngularRoutes } from '@/lib/utils';
 import { Query } from '@/services/query';
+import { DataSource, SCHEMA_NOT_SUPPORTED } from '@/services/data-source';
+import notification from '@/services/notification';
+import recordEvent from '@/services/recordEvent';
 
 import QueryPageHeader from './components/QueryPageHeader';
+import SchemaBrowser from './components/SchemaBrowser';
+
+import './query-source.less';
+
+function getSchema(dataSource, refresh = undefined) {
+  if (!dataSource) {
+    return Promise.resolve([]);
+  }
+
+  return dataSource.getSchema(refresh)
+    .then((data) => {
+      if (data.schema) {
+        return data.schema;
+      } else if (data.error.code === SCHEMA_NOT_SUPPORTED) {
+        return [];
+      }
+      return Promise.reject(new Error('Schema refresh failed.'));
+    })
+    .catch(() => {
+      notification.error('Schema refresh failed.', 'Please try again later.');
+      return Promise.resolve([]);
+    });
+}
+
+function QuerySource(props) {
+  const [query, setQuery] = useState(props.query);
+  const [dataSources, setDataSources] = useState([]);
+  const dataSource = useMemo(() => (find(dataSources, { id: query.data_source_id }) || null), [query, dataSources]);
+  const [schema, setSchema] = useState([]);
+  const refreshSchemaTokenRef = useRef(null);
+
+  useEffect(() => {
+    recordEvent('view_source', 'query', query.id);
+
+    let isCancelled = false;
+    DataSource.query().$promise.then((data) => {
```

Comment:
I was thinking if this could also go in the angular `resolve`. Universal Router supports [async routes](https://github.com/kriasoft/universal-router/blob/master/docs/api.md#async-routes) and that should be a fit.

My thoughts ended up in: if it's sth that is a "main" resource for the page that should be in there (basically when it's required for the page to work), otherwise it's better to have a "loading state". Thus, I eventually concluded that in this case it's better to render the rest of the page and have it as is, but wanted to share this anyway :stuck_out_tongue_closed_eyes: 

PS: It's better to have a loading state in the Data Source selector than showing the Data source id while it's loading.

Reply:
In this case, data source list is not a blocker for the page (there are some issues with QueryEditor component, but I'll address them later). Adding loader is a good idea :+1: